### PR TITLE
Additional changes to first class function

### PR DESCRIPTION
**Pull request by sklam**
base on #5287

Addresses items in https://github.com/numba/numba/pull/5287#issuecomment-604464489

### CODE REVIEW

File: numba/core/utils.py [hunk 1]
```diff
@@ -495,6 +493,10 @@ def unified_function_type(numba_types):
     If any of the Numba function types is a Dispatcher type, the
     unified function type will be UndefinedFunctionType instance.
     """
+    from numba.core.errors import NumbaExperimentalFeatureWarning
```
Comment (stuartarchibald) [inline numba/core/utils.py]: Move import to the top? Or is there circular import problems.

**Comment by sklam** [comment]
cuda smoke test at `numba_smoketest_cuda_42`

File: numba/core/utils.py (hunk 1 above)
Comment (sklam) [reply numba/core/utils.py]: it's a circular import problem.

**Comment by stuartarchibald** [comment]
Seems like the warning triggers a lot more often than expected?!

**Comment by sklam** [comment]
The warning needs to be moved under the guard. The `unified_function_type` function is used when it doesn't know the inputs contain function types.

**Comment by sklam** [comment]
cuda smoke test passed.

File: docs/source/reference/types.rst [hunk 2]
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
...
+========================   ==============   ============    ===========
+Context                    cfunc compiled   jit compiled    WAP objects
+========================   ==============   ============    ===========
+Can be used as arguments   yes              yes             yes
+Namespace scoping          yes              yes             no
+Can be called              yes              yes             yes
+Can be used as items       yes              yes             yes
+Automatic overload         no               yes             no
+Can be returned            yes              yes             yes
+========================   ==============   ============    ===========
```
Comment (eric-wieser) [inline docs/source/reference/types.rst]: nit: consider sorting the `no`s together for easier reading
```suggestion
========================   ============   ==============   ===========
Context                    jit compiled   cfunc compiled   WAP objects
========================   ============   ==============   ===========
Can be used as arguments   yes            yes              yes
Can be called              yes            yes              yes
Can be used as items       yes            yes              yes
Can be returned            yes            yes              yes
Namespace scoping          yes            yes              no
Automatic overload         yes            no               no
========================   ============   ==============   ===========
```
(swapped rows and columns)

Comment (pearu) [reply docs/source/reference/types.rst]: Good idea.

The table needs the following update as well:
`Namespace scoping` for `WAP` is `yes`.

File: docs/source/reference/types.rst [hunk 3]
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
...
+Wrapper Address Protocol (WAP, see below) with the following restrictions:
+
+========================   ============   ==============   ===========
+Context                    jit compiled   cfunc compiled   WAP objects
+========================   ============   ==============   ===========
+Can be used as arguments   yes            yes              yes
+Can be called              yes            yes              yes
+Can be used as items       yes            yes              yes
+Can be returned            yes            yes              yes
+Namespace scoping          yes            yes              no
```
Comment (sklam) [inline docs/source/reference/types.rst]:
```suggestion
Namespace scoping          yes            yes              yes
```

**Pull request merged**